src/metabase/lib/schema/expression.cljc

```diff
@@ -1,215 +1,136 @@
 (ns metabase.lib.schema.expression
   (:require
-   [malli.core :as mc]
+   [metabase.lib.dispatch :as lib.dispatch]
    [metabase.lib.schema.common :as common]
-   [metabase.lib.schema.filter :as filter]
-   [metabase.lib.schema.literal :as literal]
-   [metabase.lib.schema.ref :as ref]
-   [metabase.lib.schema.temporal-bucketing :as temporal-bucketing]
+   [metabase.shared.util.i18n :as i18n]
    [metabase.types]
+   [metabase.util.malli :as mu]
    [metabase.util.malli.registry :as mr]))
 
 (comment metabase.types/keep-me)
 
-(defn- ref-of-base-type [base-type]
+(defmulti type-of*
+  "Impl for [[type-of]]. Use [[type-of]], but implement [[type-of*]].
+
+  For MBQL clauses, try really hard not return an ambiguous set of possible types! Calculate things and determine what
+  the result type will be!
+
+  If we don't have enough information to determine the type (e.g. a `:field` clause that needs a metadata provider to
+  determine the type), return `::expression/type.unknown`. This is a temporary workaround until we figure out how to
+  always have type info!"
+  {:arglists '([expr])}
+  #?(:clj (fn [x]
+            ;; only for clojure! Use the class name as the dispatch type rather than `:type/*`. This is so we can
+            ;; implement support for some Clojure-only classes like `BigDecimal` or `java.time.OffsetDateTime`, for
+            ;; use inside QP code or whatever.
+            (let [dispatch-value (lib.dispatch/dispatch-value x)]
+              (if (= dispatch-value :type/*)
+                (type x)
+                dispatch-value)))
```

Comment:
This code is legit in CLJS as well - `type` works pretty much the same way. We might well want to support `moment.Moment` or `js/Date`, and it seems harmless to unify the code here even if we're not using it in CLJS right now.

Reply:
To be clear, we still need `dispatch-value` because of how some values are handled differently between the two langauges. But the basic principle of falling back to `type` and using something specific like `java.time.OffsetDateTime` or `js/Date` is sound.

Reply:
good point. Fixed